includes/classes/SupportMonitor/DBQueryMonitor.php

```diff
@@ -0,0 +1,173 @@
+
+	/**
+	 * The transient name
+	 *
+	 * This transient stores all the queries logged.
+	 */
+	const TRANSIENT_NAME = 'tenup_experience_db_queries';
+
+	/**
+	 * Setup module
+	 */
+	public function setup() {
+		$production_environment = Monitor::instance()->get_setting( 'production_environment' );
+		if ( 'no' === $production_environment || apply_filters( 'tenup_experience_log_heavy_queries', false ) ) {
+			add_filter( 'query', [ $this, 'maybe_log_query' ] );
+		}
+	}
+
+	/**
+	 * Conditionally log a query
+	 *
+	 * Get all potential heavy queries (CREATE, ALTER, etc.) and store it,
+	 * ignoring transients and options by default.
+	 *
+	 * @param string $query The SQL query.
+	 * @return string
+	 */
+	public function maybe_log_query( $query ) {
+		global $wpdb;
+
+		if ( ! preg_match( '/^\s*(create|alter|truncate|drop|insert|delete|update|replace)\s/i', $query ) ) {
+			return $query;
+		}
+
+		if ( false !== strpos( $query, 'transient_' ) ) {
+			return $query;
+		}
+
+		if ( false !== strpos( $query, $wpdb->options ) ) {
+			return $query;
+		}
+
+		if ( apply_filters( 'tenup_experience_log_query', true, $query ) ) {
+			$this->log_query( $query );
+		}
+
+		return $query;
+	}
+
+	/**
+	 * Get the logged queries. Also removes from the transient all queries logged for more than 7 days.
+	 *
+	 * @return array
+	 */
+	public function get_report() {
+		$date_limit = strtotime( '7 days ago' );
+
+		$stored_queries = array_filter(
+			(array) get_transient( self::TRANSIENT_NAME ),
+			function ( $query_date ) use ( $date_limit ) {
+				return strtotime( $query_date ) > $date_limit;
+			},
+			ARRAY_FILTER_USE_KEY
+		);
+
+		set_transient( self::TRANSIENT_NAME, $stored_queries );
+
+		return $stored_queries;
+	}
+
+	/**
+	 * Log/store a SQL query
+	 *
+	 * Queries are stored like:
+	 *
+	 *    'YYYY-MM-DD' => [
+	 *        'a5be998feee8968155052c4d332a7223' => [ // md5 of file:line:query
+	 *            'query' => 'ALTER TABLE wp_my_db_heavy_plugin CHANGE COLUMN `id` id BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT',
+	 *            'file' => '/var/www/html/wp-content/plugins/my-db-heavy-plugin/my-db-heavy-plugin.php',
+	 *            'line' => 53,
+	 *            'count' => 1,
+	 *        ]
+	 *    ]
+	 *
+	 * @param string $query The SQL query.
+	 */
+	protected function log_query( $query ) {
+		static $stored_queries;
+		if ( empty( $stored_queries ) ) {
+			$stored_queries = get_transient( self::TRANSIENT_NAME ) ?? [];
+		}
+
+		$current_date = date_i18n( 'Y-m-d' );
+
+		if ( ! isset( $stored_queries[ $current_date ] ) ) {
+			$stored_queries[ $current_date ] = [];
+		}
+
+		$main_caller = $this->find_main_caller();
+
+		$key = md5(
+			$main_caller['file'] . ':' .
+			$main_caller['line'] . ':' .
+			$query
+		);
+
+		if ( isset( $stored_queries[ $current_date ][ $key ] ) ) {
+			$stored_queries[ $current_date ][ $key ]['count']++;
+		} else {
+			$stored_queries[ $current_date ][ $key ] = [
+				'query' => stripslashes( $query ),
```

Comment:
@felipeelia How will we be making sure $query doesn't contain any sensitive information such as user e-mail addresses or other information? We'll want to be really careful about sending that sort of sensitive information to Support Monitor where we could be potentially liable for storing that information if we were ever hacked. I'd recommend we do some sort of cleanup or see if you can get the version of the query that has placeholders like %s in place of actual data.

Reply:
@bengreeley my bad on this one. Trying to get the version of the query with placeholders can be hard (or near impossible depending on the plugin), so cleaning it up would be probably easier. I'll work on that, thanks!

Reply:
@bengreeley with the change I made, I'm curious to know your opinion about this topic. We are not sending the query anymore but only storing it in a transient. As the entire feature is only available in non-production environments, do you think we still need to escape the queries? (I don't want to make it insecure but also don't want to over-engineer it). If we decide to escape them, would you be okay with using a lib like [this](https://github.com/greenlion/PHP-SQL-Parser)?

Reply:
I wonder if you create a regex or search/replace to remove anything between '' and "" if that would be sufficient? I'm not as worried if we're storing it on the server, but we'll want to be aware that if we have queries that have sensitive user information that's available to anybody or any plugin, we'll want to think of a way to clean it up.As long as it's information that is freely available in any other WP table it shouldn't be a bit deal, but should be given adequate testing.

Reply:
@bengreeley I've pushed a commit that does that. When I save a long post these are the queries being saved (the post update and the revision created):

```
Query: UPDATE `wp_posts` SET `post_author` = 1, `post_date` = ?, `post_date_gmt` = ?, `post_content` = ?, `post_content_filtered` = ?, `post_title` = ?, `post_excerpt` = ?, `post_status` = ?, `post_type` = ?, `comment_status` = ?, `ping_status` = ?, `post_password` = ?, `post_name` = ?, `to_ping` = ?, `pinged` = ?, `post_modified` = ?, `post_modified_gmt` = ?, `post_parent` = 0, `menu_order` = 0, `post_mime_type` = ?, `guid` = ? WHERE `ID` = 7
File: /var/www/html/wp-includes/post.php
Line: 4196
Count: 1

Query: INSERT INTO `wp_posts` (`post_author`, `post_date`, `post_date_gmt`, `post_content`, `post_content_filtered`, `post_title`, `post_excerpt`, `post_status`, `post_type`, `comment_status`, `ping_status`, `post_password`, `post_name`, `to_ping`, `pinged`, `post_modified`, `post_modified_gmt`, `post_parent`, `menu_order`, `post_mime_type`, `guid`) VALUES (1, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 7, 0, ?, ?)
File: /var/www/html/wp-includes/post.php
Line: 4219
Count: 1
```

Does that look okay?